src/main/scala/com/fulcrumgenomics/commons/async/AsyncIterator.scala

```diff
@@ -0,0 +1,94 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package com.fulcrumgenomics.commons.async
+
+import java.util.concurrent.{ArrayBlockingQueue, BlockingQueue, LinkedBlockingDeque}
+
+object AsyncIterator {
+  /** Builds an [[AsyncIterator]] and starts it.
+    *
+    * @param source the source iterator to consume
+    * @param bufferSize the maximum number of elements to buffer before the blocking when consuming the source, or None
+    *                   if unbounded.
+    * @tparam T the type of object to consume
+    */
+  def apply[T](source: Iterator[T], bufferSize: Option[Int] = None): AsyncIterator[T] = {
+    new AsyncIterator[T](source, bufferSize).start()
+  }
+}
+
+/** An asynchronous wrapper for an [[Iterator]].  A separate thread will be created to consume the source iterator.
+  * Will buffer up to [[bufferSize]] elements before the blocking when consuming the source.
+  *
+  * @param source the source iterator to consume
+  * @param bufferSize the maximum number of elements to buffer before the blocking when consuming the source, or None
+  *                   if unbounded.
+  * @tparam T the type of object to consume
+  */
+class AsyncIterator[T](private val source: Iterator[T], bufferSize: Option[Int] = None) extends Iterator[T] with AsyncRunnable {
+  bufferSize.foreach(size => require(size > 0, s"bufferSize must be greater than zero when given, found $size"))
+
+  private val queue: BlockingQueue[T] = bufferSize match {
+    case Some(size) => new ArrayBlockingQueue[T](size)
+    case None       => new LinkedBlockingDeque[T]()
+  }
+
+  protected def execute(): Unit = this.source.foreach(this.queue.put)
+
+  /** Returns true if there exists more elements, false otherwise */
+  def hasNext(): Boolean = {
+    checkAndRaise()
+
+    // If the queue is empty but the underlying thread is not done, then there may or may not be new
+    // items that will be put in the queue.
+    while (this.queue.isEmpty && !this.done) {
+      Thread.sleep(50)
+      checkAndRaise() // check if there was any exception in the underlying thread
+    }
+
+    // At this point, either there are items in the queue, or the underlying thread has consumed the source
+    if (!this.queue.isEmpty) true
+    else { // the queue is empty, and the source has been consumed
+      // wait for the execute method to complete, regardless of it an exception was thrown
+      awaitDone()
+      // check for an exception
+      checkAndRaise()
+      !this.queue.isEmpty
+    }
+  }
+
+  /** Gets the next item. */
+  def next(): T = {
+    checkAndRaise()
+    if (this.hasNext()) {
+      val item = tryAndModifyInterruptedException("Interrupted waiting on taking from the queue.") { this.queue.take() }
```

Comment:
You should be able to use `this.queue.remove()` here which is a non-blocking method that throws an exception if no element is available, but then you don't need to worry about `InterruptedException`s.

Reply:
Done